src/Util/SmartPaginatorFactory.php

```diff
@@ -60,13 +61,91 @@ public static function create(ProxyQueryInterface $proxyQuery, array $hints = []
         // To stay safe fetch join only when we have single primary key and joins
         $paginator = new Paginator($query, $hasSingleIdentifierName && $hasJoins);
 
-        $hasHavingPart = null !== $queryBuilder->getDQLPart('having');
-
         // it is only safe to disable output walkers for really simple queries
-        if (!$hasHavingPart && $hasSingleIdentifierName) {
+        if (self::canDisableOutPutWalkers($queryBuilder)) {
             $paginator->setUseOutputWalkers(false);
         }
 
         return $paginator;
     }
+
+    /**
+     * @see https://github.com/doctrine/orm/issues/8278#issue-705517756
+     */
+    private static function canDisableOutPutWalkers(QueryBuilder $queryBuilder): bool
+    {
+        // does not support queries using HAVING
+        if (null !== $queryBuilder->getDQLPart('having')) {
+            return false;
+        }
+
+        $fromParts = $queryBuilder->getDQLPart('from');
+
+        // does not support queries using multiple entities in FROM
+        if (1 !== \count($fromParts)) {
+            return false;
+        }
+
+        $fromPart = current($fromParts);
+
+        $classMetadata = $queryBuilder
+            ->getEntityManager()
+            ->getClassMetadata($fromPart->getFrom());
+
+        $identifierFieldNames = $classMetadata->getIdentifierFieldNames();
+
+        // does not support entities using a composite identifier
+        if (1 !== \count($identifierFieldNames)) {
+            return false;
+        }
+
+        $identifierName = current($identifierFieldNames);
+
+        // does not support entities using a foreign key as identifier
+        if ($classMetadata->hasAssociation($identifierName)) {
+            return false;
+        }
+
+        // does not support queries using a field from a toMany relation in the ORDER BY clause
+        if (self::hasOrderByWithToManyAssociation($queryBuilder)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static function hasOrderByWithToManyAssociation(QueryBuilder $queryBuilder): bool
+    {
+        $joinParts = $queryBuilder->getDQLPart('join');
+
+        if (0 === \count($joinParts)) {
+            return false;
+        }
+
+        $orderByParts = $queryBuilder->getDQLPart('orderBy');
+
+        if (0 === \count($orderByParts)) {
+            return false;
+        }
+
+        $joinAliases = [];
+
+        foreach ($joinParts as $joinPart) {
+            foreach ($joinPart as $join) {
+                $joinAliases[] = $join->getAlias();
+            }
+        }
+
+        foreach ($orderByParts as $orderByPart) {
+            foreach ($orderByPart->getParts() as $part) {
+                foreach ($joinAliases as $joinAlias) {
+                    if (0 === strpos($part, $joinAlias)) {
```

Comment:
Which values can be hold in `$part` and `$joinAlias`? My concern is about avoiding `strpos()` to return `0` in cases like `strpos('alias', 'a')`, as I think it isn't the expected behavior here.

Reply:
Interesting, thanks for pointing it out, I think it could be fixed appending a dot to the `$joinAlias` part. 

`$joinAliases` could contain any join alias (could be `a`) and `$part` contains the ordering field that could be `alias.name`.

I'll try and add a test for that.